### ISSUE TITLE
Perf-Drop (factor=2) Ubuntu-vs-Windows on same PC (dual-boot)

### ISSUE THREAD

**Issue by alex77g2** [pytorch/pytorch#110334]
### 🐛 Describe the bug

On my Dual-Boot-PC (Ubuntu + Windows) same Torch-Code (same folder on NTFS-drive) shows factor 2 runtime slowdown on Windows.
I disabled write Parquet-Files and shuffle turned off, to exclude Filesystem or other issues. I do not use the Compile-feature, afaik.
Both: torch.device("cuda"), torch.get_default_dtype=torch.float32, torch.cuda.is_available()=True
python -m cProfile -o test.cprof mnistBenchmark.py # (.cprof attached for both systems)
Overall Runtume: Windows=30s, Ubtuntu=15s, Ubuntu+cProfile=23s (cProfile does some Ubuntu-only slowdown, but this is not the issue here).

It would need some hours to cleanup the code form confidential details, so I dont want to upload it to public at the moment, but could send python files on a non-public-channel. I think the issue is of general nature as it happens with textbook-MNIST-examples and affects all solvers (at least Adam and my own new solver in the same way).

### Versions

Hardware: PC, 32GB-DDR5, Intel i5-12400, RTX-4060-GPU, fast M2-PCIe4-SSD.
Common-Software:  Python 3.11.5, CUDA-12.2-Update2, all System-Updates.
Linux: Ubuntu 22.04, NVIDIA-Linux-x86_64-535.113.01
Windows 10, 537.32 (but last 2 verions had same issue).
Simple Net-config: MNIST, 7960 parameter, Float32, Relu, Shuffle=Off, WriteToParquet=Off, 60000/256=235 steps x 10 epochs, many optimizers (e.g. Adam).

pip list # Windows 10 + Python 3.11.5 + Cuda-12.2
numpy   1.25.2
torch   2.2.0.dev20230929+cu121 # use dev for Cuda on Windows
// no nvidia- modules

pip list # Ubuntu 22.04 + Python 3.11.5 + Cuda-12.2
numpy                    1.26.0
nvidia-cublas-cu11       11.10.3.66
nvidia-cuda-runtime-cu11 11.7.99
nvidia-cudnn-cu11        8.5.0.96
torch                    2.0.1

##################################
Collecting environment information...
PyTorch version: 2.2.0.dev20230929+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Pro
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.5 (tags/v3.11.5:cce6ba9, Aug 24 2023, 14:38:34) [MSC v.1936 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.19045-SP0
Is CUDA available: True
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4060
Nvidia driver version: 537.42
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture=9
CurrentClockSpeed=2500
DeviceID=CPU0
Family=205
L2CacheSize=7680
L2CacheSpeed=
Manufacturer=GenuineIntel
MaxClockSpeed=2500
Name=12th Gen Intel(R) Core(TM) i5-12400
ProcessorType=3
Revision=

Versions of relevant libraries:
[pip3] numpy==1.25.2
[pip3] torch==2.2.0.dev20230929+cu121
[pip3] torchaudio==2.2.0.dev20230929+cu121
[pip3] torchvideo==0.0.0
[pip3] torchvision==0.17.0.dev20230929+cu121
[conda] Could not collect

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite @ptrblck

**Comment by alex77g2** [comment]
The reason seems to be:
<method 'to' of 'torch._C.TensorBase' objects> # **x 4.98 (+398%) !!**
<method 'item' of 'torch._C.TensorBase' objects> # **x 16.3 (+1500) !!!**
Runtime explodes, but numcalls is same ( dtype + shape is identical ).

python3 -m snakeviz ubuntu.cprof
// ncalls, tottime, percall, cumtime, percall, name
# Ubuntu:
34805	17.99	0.0005169	17.99	0.0005169	~:0(<method 'acquire' of '_thread.lock' objects>) // faster on Win (13 s)
20886	0.9545	4.57e-05	0.9646	4.618e-05	~:0(<method 'to' of 'torch._C._TensorBase' objects>)
11470	0.6712	5.851e-05	0.6712	5.851e-05	~:0(<built-in method torch._C._nn.linear>)
2350	0.4791	0.0002039	0.4791	0.0002039	~:0(<method 'run_backward' of 'torch._C._EngineBase' objects>)
10	0.1852	0.01852	9.41	0.941	UniversalTorchModelTrainer.py:74(train_epoch)
12943	0.1572	1.214e-05	0.1572	1.214e-05	~:0(<method 'item' of 'torch._C._TensorBase' objects>)
1	8.8e-05	8.8e-05	8.8e-05	8.8e-05	~:0(<built-in method torch._C._cuda_init>) // zero time
# Windows10:
30038	13.08	0.0004353	13.08	0.0004353	~:0(<method 'acquire' of '_thread.lock' objects>) // even 5 s faster at Win
20886	4.752	0.0002275	5.329	0.0002552	~:0(<method 'to' of 'torch._C.TensorBase' objects>) // x 4.98 (+398%) !!
12943	2.556	0.0001974	2.556	0.0001974	~:0(<method 'item' of 'torch._C.TensorBase' objects>) // x 16.3 (+1500) !!!
11470	1.603	0.0001397	1.603	0.0001397	~:0(<built-in method torch._C._nn.linear>) // x 2.39 (+139%)
2350	0.7137	0.0003037	0.7137	0.0003037	~:0(<method 'run_backward' of 'torch._C._EngineBase' objects>) // +48%
1	0.576	0.576	0.576	0.576	~:0(<built-in method torch._C._cuda_init>) // + 576 ms (vs < 1ms)

**Comment by alex77g2** [comment]
[pip_list_lin.txt](https://github.com/pytorch/pytorch/files/12774954/pip_list_lin.txt)
[pip_list_win.txt](https://github.com/pytorch/pytorch/files/12774955/pip_list_win.txt)
[cprof_2x.zip](https://github.com/pytorch/pytorch/files/12774957/cprof_2x.zip)

**Comment by tringwald** [comment]
You seem to be using two different versions of `torch` on Ubuntu and Windows. Can you repeat the profiling with Ubuntu also using `torch==2.2.0.dev20230929+cu121`?

**Comment by alex77g2** [comment]
Yes, but I want to point out that something potentially broke in the new dev-version.
The previous windows-versions did not detect CUDA (avail=false). Whatever was the default version on windows (ignoring dev) did not allow for CUDA usage. So I have to repeat it on Ubuntu-side, which will take some hours.
In between perhaps someone has any idea why DeviceToDeviceCopy could show significant changes between Windows and Ubuntu ?

**Comment by alex77g2** [comment]
> You seem to be using two different versions of `torch` on Ubuntu and Windows. Can you repeat the profiling with Ubuntu also using `torch==2.2.0.dev20230929+cu121` and/or Windows also using `2.0.1`?

[cprof_3x.zip](https://github.com/pytorch/pytorch/files/12778188/cprof_3x.zip)
Now Profiling for same torch-2.2.0.dev20231001+cu121.
3x Profiles: 1x Win (2.2.0dev), 1x Ubuntu (2.2.0dev), 1x Ubuntu (2.0.1+cu117).
I changed my code to reduce CopyToCudaDevice (which was good for both systems), but still much aster on Ubuntu.

**Comment by alex77g2** [comment]
I wrote a trivial benchmark that shows the issue within 10 lines of python.
[torch_bench.zip](https://github.com/pytorch/pytorch/files/12778199/torch_bench.zip)
`
    dt = time_time()
    for i in range(cyc):
        v1 = torch.ones([1024], dtype=torch.int32)
        v1 = v1.to(device)
    dt = time_time() - dt
`
# Windows 10
2.2.0.dev20231001+cu121 torch.float32 True False
TO(4B) = 4774.541 ms / 100000, 11.936 us / byte
TO(4K) = 349.001 ms / 10000, 8.725 us / KB
TO(4M) = 57.001 ms / 100, 0.139 us / KB
TO(1G) = 2285.999 ms / 10, 0.218 us / KB

# Ubuntu 22.04
2.2.0.dev20231001+cu121 torch.float32 True False
TO(4B) = 735.224 ms / 100000, 1.838 us / byte
TO(4K) = 75.294 ms / 10000, 1.882 us / KB
TO(4M) = 43.819 ms / 100, 0.107 us / KB
TO(1G) = 1742.462 ms / 10, 0.166 us / KB

# Result / Difference
CopyToCudaDevice .to(cuda) takes 4x on Windows for 4 Byte and 4 KB transfers.
4MB and 1 GB shows only 20% slowdown on Windows.
There seems to be a constant delay per call, which hurts more (relative ratio) below 1 MB.

**Comment by alex77g2** [comment]
> You seem to be using two different versions of `torch` on Ubuntu and Windows. Can you repeat the profiling with Ubuntu also using `torch==2.2.0.dev20230929+cu121` and/or Windows also using `2.0.1`?

How to proceed? Is this an Nvidia-driver, Microsoft-Windows, or Torch-Issue ? If you could exclude something strange in Torch-implementation in torch.to(device), this would help much.

**Comment by taras-janea** [comment]
@alex77g2 most probably PyTorch is limited by the transfer bandwidth. In order to verify it, could you please download [bandwidthtest.cu](https://github.com/NVIDIA-developer-blog/code-samples/blob/master/series/cuda-cpp/optimize-data-transfers/bandwidthtest.cu) for measuring the transfer bandwidth, compile and run it on both machines:
1. By default, transfer size is set to 16 MB, it can be adjusted by changing `nElements`;
2. `nvcc bandwidthtest.cu -o bandwidthtest`
3. ./bandwidthtest

The expected output (with default size) should be similar to:
```
Device: GeForce GTX 680
Transfer size (MB): 16

Pageable transfers
  Host to Device bandwidth (GB/s): 5.368503
  Device to Host bandwidth (GB/s): 5.627219

Pinned transfers
  Host to Device bandwidth (GB/s): 6.186581
  Device to Host bandwidth (GB/s): 6.670246
```

If you see a correlation with your previous results, then PyTorch is limited by the transfer bandwidth. If not, then could you please update the PyTorch to the latest version, and provide:
1. Output of [torch_bench.zip](https://github.com/pytorch/pytorch/files/12778199/torch_bench.zip)
2. Output of the [bandwidthtest.cu](https://github.com/NVIDIA-developer-blog/code-samples/blob/master/series/cuda-cpp/optimize-data-transfers/bandwidthtest.cu) for the interested size;
3. Output of [script](https://raw.githubusercontent.com/pytorch/pytorch/main/torch/utils/collect_env.py);

**Comment by taras-janea** [comment]
Closing this issue due to inactivity. If you continue to experience this problem or have any related concerns, please don't hesitate to open a new issue.
